app/domain/authentication/authn_k8s/application_identity.rb

```diff
@@ -140,9 +142,18 @@ def prefixed_k8s_annotations prefix
         @host_annotations.select do |a|
           annotation_name = a.values[:name]
 
+          # Calculate the granularity level of the annotation.
+          # For example, the annotation "authn-k8s/namespace" is in the general level,
+          # and applies to every host that tries to authenticate with the k8s authenticator, regardless
+          # of the service id.
+          # The annotation "authn-k8s/#{@service_id}/namespace" is on the service-id level,
+          # and applies only to hosts trying to authenticate with the authenticator "authn-k8s/#{@service_id}".
+          annotation_granularity_level = annotation_name.split('/').length
+          prefix_granularity_level = prefix.split('/').length
+
```

Comment:
Much better.

I have two comments, one important and one cosmetic, neither of them requiring you to make any changes atm.

## Important comment

What is going on here is indicating that we are missing an `AnnotationName` value object (or something similar) which would house the logic and allow you to write something like:
```ruby
annotation_name.granularity == prefix.granularity
```
To be clear, the details of the correct interface are something I'd need more context to advise on, I'm just trying to indicate the general idea that you don't want code like the above responsible for having to deal with this kind of thing.  More importantly, without a value object to house it, the same logic will often spread out in multiple places in your code base.

This has been a huge problem with the details around constructing conjur variables, eg, you'll see colon-joining code scattered in tens of places throughout this code base and even others.  We've made this mistake in other places too.

This change is probably out of scope for this PR, I'm just bringing it up as something to be aware of when you see it and in future design work.

## Cosmetic comment

Turn on a paragraph hard-wrapping hotkey in your editor so your comments will have clean 80-char margins :)

Reply:
Thanks, we'll take this as part of the refactor we'll do for the application identity.